# Project description  
The coronavirus took the entire world by surprise, changing everyone's daily routine. City dwellers no longer spent their free time outside, going to cafes and malls; more people were at home, reading books. That attracted the attention of startups that rushed to develop new apps for book lovers. 

You've been given a database of one of the services competing in this market. It contains data on books, publishers, authors, and customer ratings and reviews of books. This information will be used to generate a value proposition for a new product.

### Task

- Find the number of user reviews and the average rating for each book.
- Identify the publisher that has released the greatest number of books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis).
- Identify the author with the highest average book rating (look only at books with at least 50 ratings).
- Find the average number of text reviews among users who rated more than 50 books.
- Find the number of books released after January 1, 2000.

### Table of Contents: <a class="anchor" id="contents"></a>

#### [1. Preparing environment and uploading data](#chapter1)
   * [1.1 Importing packages](#chapter2)
   * [1.2 Database connection](#chapter3)
   * [1.3 Checking the tables](#chapter11)

#### [2. Styding the data](#chapter4)
   * [2.1 Finding the number of user reviews and the average rating for each book](#chapter5)
   * [2.2 Identifying the publisher with greatest number of books (50+ pages)](#chapter6)
   * [2.3 Identifying the author with the highest average book rating](#chapter7)
   * [2.4 Finding the average number of text reviews among users who rated 50+ books](#chapter8)
   * [2.5 Finding the number of books released after January 1, 2000](#chapter9)

### 1.  Preparing environment and uploading data <a class="anchor" id="chapter1"></a>

#### 1.1 Importing packages <a class="anchor" id="chapter2"></a>

In [1]:
import pandas as pd
from sqlalchemy import create_engine

[Back to the Table of Contents](#contents)

#### 1.2 Database connection <a class="anchor" id="chapter3"></a>

In [2]:
#connecting to the Data Base
db_config = {'user': 'test',         # user name
             'pwd': 'test', # password
             'host': 'test',
             'port': 6432,              # connection port
             'db': 'test'}          # the name of the data base

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

The database connection is stored in the engine variable. 

[Back to the Table of Contents](#contents)

#### 1.3 Checking the tables <a class="anchor" id="chapter11"></a>

Connectiont to the database was created.  
Now we need to write code to download data from the tables to the csv files, then we can open files through pandas and store them into data frame to studying the tables.  
Let's run an SQL query:

In [3]:
#creating query
query = '''SELECT * FROM books'''
query_1 = '''SELECT * FROM authors'''
query_2 = '''SELECT * FROM publishers'''
query_3 = '''SELECT * FROM ratings'''
query_4 = '''SELECT * FROM reviews'''
#downloading data from the database
books = pd.io.sql.read_sql(query, con = engine)
authors = pd.io.sql.read_sql(query_1, con = engine)
publishers = pd.io.sql.read_sql(query_2, con = engine)
ratings = pd.io.sql.read_sql(query_3, con = engine)
reviews = pd.io.sql.read_sql(query_4, con = engine)

Let's download data from the each table to csv files (I am saving scv for myself):

In [4]:
books.to_csv('books.csv', index = False)
authors.to_csv('authors.csv', index = False)
publishers.to_csv('publishers.csv', index = False)
ratings.to_csv('ratings.csv', index = False)
reviews.to_csv('reviews.csv', index = False)

Now we can open csv files to check the tables or we can use parameters with downloaded data:

In [5]:
#reading csv files
df_books = pd.read_csv('books.csv')
df_authors = pd.read_csv('authors.csv')
df_publishers = pd.read_csv('publishers.csv')
df_ratings = pd.read_csv('ratings.csv')
df_reviews = pd.read_csv('reviews.csv')

Let's print the first rows of each table:

In [6]:
#from csv
df_books.head()
#OR from parameter with df
books.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

Table books contains data on books:

- `book_id`
- `author_id`
- `title`
- `num_pages` — number of pages
- `publication_date`
- `publisher_id`

In [7]:
authors.head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

Table authors contains data on authors:

- `author_id`
- `author`

In [8]:
publishers.head()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

Table publishers contains data on publishers:

- `publisher_id`
- `publisher`

In [9]:
ratings.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

Table ratings contains data on user ratings:

- `rating_id`
- `book_id`
- `username` — the name of the user who rated the book
- `rating`

In [10]:
reviews.head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

Table reviews contains data on customer reviews:

- `review_id`
- `book_id`
- `username` — the name of the user who reviewed the book
- `text` — the text of the review

[Back to the Table of Contents](#contents)

### 2. Styding the data <a class="anchor" id="chapter4"></a>

#### 2.1 Finding the number of user reviews and the average rating for each book  <a class="anchor" id="chapter5"></a>

In [11]:
#creating query
query_task_1 = '''
SELECT
    books.book_id,
    COUNT(reviews.review_id) AS review_amount,
    AVG(ratings.rating) AS avg_rating
FROM
    books
LEFT JOIN 
    ratings on ratings.book_id = books.book_id    
LEFT JOIN 
    reviews on reviews.book_id = books.book_id
GROUP BY
    books.book_id;
    '''
#downloading data from the database
task_1 = pd.io.sql.read_sql(query_task_1, con = engine)
#checking result
task_1

book_id  review_amount  avg_rating
0        652              4    4.500000
1        273              4    4.500000
2         51             60    4.250000
3        951              4    4.000000
4        839             28    4.285714
..       ...            ...         ...
995       64             52    4.230769
996       55              4    5.000000
997      148             21    3.428571
998      790              4    3.500000
999      828              4    3.000000

[1000 rows x 3 columns]

Table shows the number of user reviews and the average rating for each book.

[Back to the Table of Contents](#contents)

####  2.2 Identifying the publisher with greatest number of books (50+ pages) <a class="anchor" id="chapter6"></a>

In [12]:
#creating query
query_task_2 = '''
SELECT
    publishers.publisher,
    COUNT(books.book_id) AS books_amount
FROM
    books
INNER JOIN 
    publishers on publishers.publisher_id = books.publisher_id
WHERE
    books.num_pages > 50
GROUP BY
    publishers.publisher
ORDER BY
    books_amount DESC
LIMIT
1;    
    '''
#downloading data from the database
task_2 = pd.io.sql.read_sql(query_task_2, con = engine)
task_2

publisher  books_amount
0  Penguin Books            42

The publisher that has released the greatest number of books were identified: `Penguin Books`.  
I chose books with more than 50 pages to exclude brochures and similar publications.

[Back to the Table of Contents](#contents)

#### 2.3 Identifying the author with the highest average book rating  <a class="anchor" id="chapter7"></a>

In [13]:
query_task_3 = '''
SELECT
    authors.author,
    AVG(ratings.rating) AS average_rating
FROM (
    SELECT
        ratings.book_id,
        books.author_id
    FROM
        ratings
    LEFT JOIN books on books.book_id = ratings.book_id
    GROUP BY
        ratings.book_id,
        books.author_id
    HAVING
        COUNT(ratings.rating_id) >=50) AS books_raited_50_times
INNER JOIN 
    ratings on ratings.book_id = books_raited_50_times.book_id
INNER JOIN 
    authors on authors.author_id = books_raited_50_times.author_id
GROUP BY
    authors.author
ORDER BY
    average_rating DESC
LIMIT
    1;
    '''
#downloading data from the database
task_3 = pd.io.sql.read_sql(query_task_3, con = engine)
#checking result
task_3

author  average_rating
0  J.K. Rowling/Mary GrandPré        4.287097

The author with the highest average book rating is `J.K. Rowling/Mary GrandPré` with average rating 4.29.  
I looked only at books rated at least 50 times.

[Back to the Table of Contents](#contents)

#### 2.4 Finding the average number of text reviews among users who rated 50+ books  <a class="anchor" id="chapter8"></a>

In [14]:
#creating query
query_task_4 = '''
SELECT 
    AVG(reviews_amount) AS avg_n_reviews
FROM 
    (SELECT
        reviews.username,
        COUNT(review_id) AS reviews_amount
    FROM 
        (SELECT 
            username
        FROM 
            ratings
        GROUP BY
            username
        HAVING
            COUNT(rating_id) > 50) AS rated_more_50
    LEFT JOIN 
        reviews on reviews.username = rated_more_50.username
    GROUP BY
        reviews.username) AS n_reviews;
    '''
#downloading data from the database
task_4 = pd.io.sql.read_sql(query_task_4, con = engine)
#checking result
task_4

avg_n_reviews
0      24.333333

The average number of text reviews among users who rated more than 50 books is 24.3.

[Back to the Table of Contents](#contents)

#### 2.5 Finding the number of books released after January 1, 2000  <a class="anchor" id="chapter9"></a>

In [15]:
#creating query
query_task_5 = '''
SELECT
    COUNT(book_id) AS books_amount
FROM
    books
WHERE
    publication_date > '2000-01-01';
    '''
#downloading data from the database
task_5 = pd.io.sql.read_sql(query_task_5, con = engine)
#checking result
task_5

books_amount
0           819

We have 819 books released after January 1, 2000.

[Back to the Table of Contents](#contents)